File: 2026/April/LC-3418-Maximum-Amount-of-Money-Robot-Can-Earn/solution.cpp
// Problem link: https://leetcode.com/problems/maximum-amount-of-money-robot-can-earn/description/

#include<vector>
using namespace std;
class Solution {
    private:
        int dp[501][501][3];
    public:
        int solve(vector<vector<int>>& grid, int i, int j, int neu) {
            int rows = grid.size();
            int cols = grid[0].size();
    
            //reached destination
            if (i == rows - 1 && j == cols - 1) {
                if (grid[i][j] < 0 && neu > 0) {
                    return 0; // neutralize the robber
                }
                return grid[i][j];
            }
    
            // Out of bounds
            if (i >= rows || j >= cols) {
                return INT_MIN;
            }
    
            // Already computed
            if (dp[i][j][neu] != INT_MIN) {
                return dp[i][j][neu];
            }
    
            //Take the current cell value
            int moveDown  = solve(grid, i + 1, j, neu);
            int moveRight = solve(grid, i, j + 1, neu);
    
            int take = grid[i][j] + max(moveDown, moveRight);
    
            //Neutralize (only if negative and we have chances left)
            int skip = INT_MIN;
            if (grid[i][j] < 0 && neu > 0) {
                int skipDown  = solve(grid, i + 1, j, neu - 1);
                int skipRight = solve(grid, i, j + 1, neu - 1);
    
                skip = max(skipDown, skipRight);
            }
    
            // Store and return best result
            return dp[i][j][neu] = max(take, skip);
        }
    
        int maximumAmount(vector<vector<int>>& coins) {
            for (int i = 0; i < 501; i++) {
                for (int j = 0; j < 501; j++) {
                    for (int k = 0; k < 3; k++) {
                        dp[i][j][k] = INT_MIN;
                    }
                }
            }
    
            return solve(coins, 0, 0, 2);
        }
    };

File: 2026/April/LC-3418-Maximum-Amount-of-Money-Robot-Can-Earn/solution.py
# Problem Link: https://leetcode.com/problems/maximum-amount-of-money-robot-can-earn/description/

from typing import List
import math
class Solution:
    def __init__(self):
        self.m  = 0
        self.n  = 0
        self.dp = []
    def __solve(self, coins, i: int, j: int, k: int) -> float:
        if i >= self.m or j >= self.n:
            return -math.inf
        if i == self.m-1 and j == self.n-1:
            if coins[i][j] < 0 and k > 0:
                return 0
            return coins[i][j]
        if self.dp[i][j][k] != -math.inf:
            return self.dp[i][j][k]
        
        down  = self.__solve(coins, i + 1, j, k)
        right = self.__solve(coins, i, j + 1, k)
        take  = coins[i][j] + max(down, right)

        skip = -math.inf
        if coins[i][j] < 0 and k > 0:
            skip_down  = self.__solve(coins, i + 1, j, k - 1)
            skip_right = self.__solve(coins, i, j + 1, k - 1)
            skip = max(skip_down, skip_right)
        
        self.dp[i][j][k] = max(take, skip)
        return self.dp[i][j][k]
    def maximumAmount(self, coins: List[List[int]]) -> int:
        self.m = len(coins)
        self.n = len(coins[0])
        self.dp = [[[-math.inf]*3 for _ in range(self.n)] for _ in range(self.m)]
        return int(self.__solve(coins, 0, 0, 2))